Client/Frontend/Browser/BrowserViewController.swift

```diff
@@ -2818,6 +2829,125 @@ extension BrowserViewController: ContextMenuHelperDelegate {
     }
 }
 
+/**
+ A third party search engine Browser extension
+**/
+extension BrowserViewController {
+
+    func addCustomSearchButtonToWebView(webView: WKWebView) {
+        //check if the search engine has already been added.
+        let domain = webView.URL?.domainURL().host
+        let matches = self.profile.searchEngines.orderedEngines.filter({$0.shortName == domain})
+        if !matches.isEmpty {
+            self.customSearchEngineButton.tintColor = UIColor.grayColor()
+            self.customSearchEngineButton.userInteractionEnabled = false
+        } else {
+            self.customSearchEngineButton.tintColor = UIConstants.SystemBlueColor
+            self.customSearchEngineButton.userInteractionEnabled = true
+        }
+
+        /*
+         This is how we access hidden views in the WKContentView
+         Using the public headers we can find the keyboard accessoryView which is not usually available.
+         Specific values here are from the WKContentView headers.
+         https://github.com/JaviSoto/iOS9-Runtime-Headers/blob/master/Frameworks/WebKit.framework/WKContentView.h
+        */
+        guard let webContentView = UIView.findSubViewWithFirstResponder(webView),
+            let input = webContentView.performSelector(Selector("inputAccessoryView")),
+            let inputView = input.takeUnretainedValue() as? UIInputView,
+            let nextButton = inputView.valueForKey("_nextItem") as? UIBarButtonItem,
+            let nextButtonView = nextButton.valueForKey("view") as? UIView else {
+                //failed to find the inputView instead lets use the findInPage bar
+                addCustomSearchButtonToFindInPage()
+                return
+            }
+
+            inputView.addSubview(self.customSearchEngineButton)
+            self.customSearchEngineButton.snp_remakeConstraints { make in
+                make.leading.equalTo(nextButtonView.snp_trailing).offset(20)
+                make.width.equalTo(inputView.snp_height)
+                make.top.equalTo(nextButtonView.snp_top)
+                make.height.equalTo(inputView.snp_height)
+            }
+    }
+
+    /**
+     This adds the customSearchButton to the findInPage container
+     for cases where the inputAccessoryView could not be found for example
+     on the iPad where it does not exist.
+     **/
+    func addCustomSearchButtonToFindInPage() {
```

Comment:
Unfortunately, the find-in-page layout (and BVC layouts in general) are very delicate, so this hack of adding the search button to the find-in-page bar might not work well.

Some steps to break it:
 1. Tap a text field to make the bar appear with the "add search" button.
 1. Tap the text field again to make the Find in Page option appear.
 1. Tap Find in Page.
 1. Tap anywhere outside the text field to make it lose focus.

After these steps, a black gap will be left at the bottom of the page where the find-in-page bar was.